sorted_array/map fast lookup

gswirski opened the pull request:
## Describe your changes


## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Commits are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.
- Updated `CHANGELOG.md'

gswirski commented:
@adr1anh this PR is a result of a comment here: https://github.com/0xMiden/miden-vm/pull/2048#discussion_r2261250290

I know @bobbinth wanted to land SMT access in September - this would be a prerequisite/an important optimization for that work. Depending on how fast we can move with #1582 , we could merge this first to unblock #2048 (once I add support for sorted key-values) and refactor later? What do you think?

bobbinth reviewed (commented):
Looks good! Thank you! Not a full review yet, but I left some comments inline.

> what is the policy for code duplication in MASM? `sorted_array::find_word` and `sorted_map::find_key` are extremely similar (they differ in stride value, find_key needs to do an additional alignment check) but I'm not sure if we prefer precise cycles optimization, or smaller code size

I think given that these procedures are not too big, duplication is fine. But also, if there is away to find how to avoid it without complicating things too much, that, of course be better.

> strictly speaking, we only need `sorted_map` version for smt::get/set methods, but I thought `sorted_array` would be generally useful for others. Can remove it if we don't want to pollute stdlib too much (which would also solve the above dillema).

I think having both is good - though, I'm not too sure about naming. Both of these are technically arrays, they just different in the type of entries. So, maybe we should just have a single `sorted_array` module with two procedures. Though, I haven't thought about the naming of these procedures too much.

> the `start_ptr, end_ptr` in the result are there because they tend to be useful until the end of the algorithm. should we drop them before returning, or let callee decide if they want to clean them up?

No strong preference here. I think more recently we started dropping the inputs and then if the caller wants them, then can always `dup` them before invoking the procedure, but there are plenty of examples where we return the inputs as well.

bobbinth commented on assembly-syntax/src/ast/instruction/advice.rs:
```diff
@@ -15,6 +15,8 @@ pub enum SystemEventNode {
     PushU64Div,
     PushFalconDiv,
     PushSmtPeek,
+    PushArrayLowerBound,
+    PushMapLowerBound,
```
I'm not sure about this naming here: I think `PushArrayLowerBound` is fine, but maybe instead of `PushMapLowerBound` it should be `PushKeyValueLowerBound`?

bobbinth reviewed (commented):
Looks good! Thank you! I did a high-level review (i.e., didn't review MASM line-by-line) - and I think this is the right structure. I left one small comment inline about naming.

> I've just merged #2145 where we moved some system events to the standard library. These also fit the pattern so we should try adapt this PR in the same way. Let me know if you have any questions around the new structure, I'd be happy to help out!

I'm thinking that this functionality is generally useful, and so, maybe we should keep these as system events. We should still though merge the latest `next` into this PR to make sure everything still works fine.

> **One thing I'm not sure about is the binary search.** I suspect we're not going to deal with large buffers in MASM. We could replace binary search with linear search to detect unsorted arrays passed to push_xxx_lowerbound and avoid a potential footgun.

Very good point! Yes, let's switch to linear scan and maybe even go as far as to scan the whole specified memory range to ensure that all entries are in increasing order. As you've mentioned, cost-wise, this should be negligible, but could save many hours on debugging.

adr1anh commented:
> I'm thinking that this functionality is generally useful, and so, maybe we should keep these as system events.

I agree it's useful, but the utility comes from both the handler and the MASM wrapper being paired together. If I wanted to use this new handler without importing the stdlib, I'd have to pretty much reimplement the wrapping procedure. With the fairly large number of checks happening here, it's really not something easy to get right.

If this functionality is generally useful then it should be available to use in a safe way, which means calling the handler through a reliable(/audited) way. 

Finally, if some power users really needed to squeeze some extra performance out of their program, they would probably want to optimize the handler as well as the wrapper, so they'd just copy both and include them as a pair in their own library.

adr1anh reviewed (commented):
Looks good! Thanks for updating to the new event structure. Left a couple of small nits.

bobbinth commented on stdlib/src/handlers/sorted_array.rs:
```diff
@@ -0,0 +1,158 @@
+
+/// Reads (start_addr, end_addr) tuple from the specified elements of the operand stack (
+/// without modifying the state of the stack), and verifies that memory range is valid.
+fn get_mem_addr_range(
```
Yes, let's put it somewhere [here](https://github.com/0xMiden/miden-vm/blob/69ebe68851e7bc3179fd8499c966a02a951744de/stdlib/src/handlers/mod.rs#L10).

gswirski commented:
All comments addressed, all features implemented, docs (including cycle counts) provided. This is now ready for another round of reviews.

Al-Kindi-0 commented on stdlib/asm/collections/sorted_array.masm:
```diff
@@ -0,0 +1,279 @@
+                #=> [is_value_found = 0, value_ptr = start_ptr, start_ptr, end_ptr]
+            else
+                # the value was not found, maybe_value_ptr = start_ptr, and the list is not empty
+                # assert that the value at start_ptr > VALUE
```
nit: assert that the value at start_ptr is greater than VALUE

gswirski replied:
@Al-Kindi-0 I find English (and natural language in general) to be extremely annoying to describe logical relationships leading to verbose descriptions and ambiguity. How would you feel about:

```
# assert that "value at start_ptr" > VALUE
```

or even more pedantic:

```
# assert that `"value at start_ptr" > VALUE`
```

or maybe:

```
# assert that `value_at_start_ptr > VALUE`
```

In my mind ">" and ">=" leave no room for interpretation vs in English it's useful to say "strictly greater" rather than just "greater".

Logical operators are shorter, faster to scan, leave no room for interpretation, and are easier to spot inconsistencies with the actual code implementation. English looks nicer, but the intended audience for these comments are other engineers, auditors and people who probably prefer denser descriptions.

Same consideration for all related comments. In my mind quoting all formulas (preferably with `) makes a ton of sense. English - not so much.

Al-Kindi-0 commented on stdlib/asm/collections/sorted_array.masm:
```diff
@@ -0,0 +1,279 @@
+            if.true
+                # the value was not found, maybe_value_ptr = end_ptr
+                # the list is not empty, otherwise maybe_value_ptr = start_ptr and handled above
+                # assert that the value before end_ptr < VALUE
```
nit: assert that the value before end_ptr is strictly less than VALUE

Al-Kindi-0 commented on stdlib/asm/collections/sorted_array.masm:
```diff
@@ -0,0 +1,279 @@
+                #=> [is_value_found = 0, maybe_value_ptr = end_ptr, start_ptr, end_ptr]
+            else
+                # The value was not found, and the maybe_value_ptr is neither the start nor the end
+                # of the list. Make sure that the value at maybe_value_ptr-1 < VALUE and at maybe_value_ptr > VALUE
```
nit: same comment here and would probably write them as two bullet points i.e., 

... make sure that 
1. the value at maybe_value_ptr-1 is strictly less than VALUE, and that
2. the value at maybe_value_ptr is strictly greater than VALUE

gswirski replied:
Agreed about bullet points. See my comment about logical relationships: https://github.com/0xMiden/miden-vm/pull/2114#discussion_r2348341757

Al-Kindi-0 commented on stdlib/asm/collections/sorted_array.masm:
```diff
@@ -0,0 +1,279 @@
+        else
+            if.true
+                # the value was not found, maybe_value_ptr = end_ptr
+                # the list is not empty, otherwise maybe_value_ptr = start_ptr and handled above
```
nit: either exchange = for "equals" or put the equation between quotation marks

Al-Kindi-0 reviewed (approved):
Looks great!
Left a few nits, but feel free to ignore

gswirski commented on stdlib/tests/collections/sorted_array.rs:
```diff
@@ -0,0 +1,218 @@
+        );
+
+        let program = build_test!(source, &[]);
+        println!("testing {key}");
```
I left it intentionally. By default `cargo test` does not show stdout of successful tests. When the test fails, you'll see "testing A,B,C,D" and you'll know which scenario caused trouble.

I don't feel strongly either way, can remove it.

PhilippGackstatter replied:
Ah, well I don't work so much in this code base so I don't have strong opinions. But I often use `--no-capture` when trying something and then the output would be interleaved by unrelated `println`s that I didn't add, so I usually avoid them to keep things cleaner.

gswirski commented:
Addressed all lints, rustdoc errors, comments and nits except for:
1. `println!("testing {key}")`
2. Replacing logical operators with their English descriptions

I think in both instances the current version provides additional value but I don't feel very strongly either way.

gswirski commented on stdlib/src/handlers/sorted_array.rs:
```diff
@@ -0,0 +1,192 @@
+    Ok(vec![AdviceMutation::extend_stack(vec![
+        Felt::from(was_key_found),
+        Felt::from(result.unwrap_or(end_addr)),
+    ])])
```
Not true. Advice stack is in reverse order than the rest of the MASM code. `adv_push.2` flips it back to the proper order.

Al-Kindi-0 commented on processor/src/operations/sys_ops/sys_event_handlers.rs:
```diff
@@ -612,6 +614,108 @@ fn push_smtpeek_result(
 }
 
+/// Pushes onto the advice stack the first pointer in [start_ptr, end_ptr) such that
+/// mem[word_ptr] >= KEY in lexicographic order of words. If all words are < KEY, returns end_ptr.
```
Missed the equality sign, thanks for explaining

gswirski commented:
> @gswirski , is this good to go?

I think so!

I addressed all the feedback except for 2 small things I didn’t agree with (but don’t feel strongly about and can easily update): https://github.com/0xMiden/miden-vm/pull/2114/#issuecomment-3291338504

bobbinth commented on processor/src/lib.rs:
```diff
@@ -950,3 +950,27 @@ pub(crate) fn add_error_ctx_to_external_error(
+
+/// Reads (start_addr, end_addr) tuple from the specified elements of the operand stack (
+/// without modifying the state of the stack), and verifies that memory range is valid.
+pub fn get_mem_addr_range(
+    process: &ProcessState,
+    start_idx: usize,
+    end_idx: usize,
+) -> Result<(u32, u32), MemoryError> {
```
A potential small improvement: maybe this could be a method on the `ProcessState`? It could look something like:
```rust
impl ProcessState {
    pub fn get_mem_addr_range(
        &self,
        start_idx: usize,
        end_idx: usize,
    ) -> Result<Range<MemoryAddress>, MemoryError> {
        ...
    }
}
```
In the above, I've also used `Range<MemoryAddress>` instead of `(u32, u32)` for the return value as I think it is a bit more self-describing (and type-safe).

gswirski replied:
Created a followup PR here: https://github.com/0xMiden/miden-vm/pull/2194

bobbinth commented on stdlib/src/handlers/sorted_array.rs:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    // Validate the entire array is non-decreasing and find the first element where `element >= key`
+    let mut current_addr = start_addr + stride;
+    while current_addr < end_addr {
```
nit: a slightly more expressive way to write this could be:
```rust
for addr in (start_addr..end_addr).step_by(stride).skip(1) {
    ...
}
```
